app/controllers/sessions_controller.rb

```diff
@@ -0,0 +1,21 @@
+# Manages the login functionality for users
+class SessionsController < ApplicationController
+
+  def new
+  end
+
+  def create
+    user = User.find_by_email(params[:email])
+    if user.try(:authenticate, params[:password])
+      session[:user_id] = user.id
+      redirect_to admin_url
```

Comment:
I added this in. However, I wasn't able to get the 'request.referrer' to work as I'd expect and I'm not sure why. It seemed to be never set.

Instead I went the route of setting the redirect during the authorization failure. Let me know what you think and if you have any suggestions.